lib/modem_key_mgmt/modem_key_mgmt.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <bsd_limits.h>
+#include <modem_key_mgmt.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define MODEM_KEY_MGMT_OP_LS "AT%CMNG=1"
+#define MODEM_KEY_MGMT_OP_RD "AT%CMNG=2"
+#define MODEM_KEY_MGMT_OP_WR "AT%CMNG=0"
+#define MODEM_KEY_MGMT_OP_RM "AT%CMNG=3"
+
+#define AT_CMNG_PARAMS_COUNT 5
+#define AT_CMNG_CONTENT_INDEX 4
+
+static char scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	int result;
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	result = at_cmd_write("AT+CMEE?", response, sizeof(response), NULL);
+	if (result == 0) {
+		if (strchr(&response[6], '1')) {
+			return 1;
+		} else {
+			return 0;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	return at_cmd_write(at_cmee_strings[(int)enable], NULL, 0, NULL);
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
+static int write_at_cmd_with_cme_enabled(char *cmd, char *buf, size_t buf_len,
+					 enum at_cmd_state *state)
+{
+	int retval;
+	int cmee_was_active = cmee_active();
+
+	if (cmee_was_active < 0) {
+		return -EFAULT;
+	}
+
+	if (!cmee_was_active) {
+		if (cmee_set(true)) {
+			return -EIO;
+		}
+	}
+
+	retval = at_cmd_write(cmd, buf, buf_len, state);
+
+	if (!cmee_was_active) {
+		if (cmee_set(false)) {
+			return -EIO;
+		}
+	}
+
+	return retval;
+}
+
+int modem_key_mgmt_write(nrf_sec_tag_t sec_tag,
+			 enum modem_key_mgnt_cred_type_t cred_type,
+			 const u8_t *buffer, u16_t buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	written = snprintf(scratch_buf, sizeof(scratch_buf), "%s,%u,%u,\"",
+			   MODEM_KEY_MGMT_OP_WR, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+	if ((written < 0) || (written >= sizeof(scratch_buf))) {
+		return -ENOBUFS;
+	}
+
+	if ((written + buffer_len + sizeof("\"\r\n")) > sizeof(scratch_buf)) {
+		return -ENOBUFS;
+	}
+
+	memcpy(&scratch_buf[written], buffer, buffer_len);
+	written += buffer_len;
+
+	memcpy(&scratch_buf[written], "\"\r\n", sizeof("\"\r\n"));
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, NULL, 0, &state);
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
```

Comment:
My comment on line 63 would make it possible to `return translate_error` without additional logic.